common/core/desktop/tests/unit/kmnkbd/test_kmx_xstring.cpp

```diff
@@ -1208,6 +1208,27 @@ void test_decxstr() {
   assert(q == p + 2);
 }
 
+void test_xstrlen() {
+  assert(xstrlen((PKMX_WCHAR)u"") == 0);
+  assert(xstrlen((PKMX_WCHAR)u"1") == 1);
+  assert(xstrlen((PKMX_WCHAR)u"1234567890") == 10);
+  assert(xstrlen((PKMX_WCHAR)u"\uFFFF\u0008\u0001a") == 2);
+  assert(xstrlen((PKMX_WCHAR)u"\uFFFF\u0014\u0001\u0002\u0003a") == 2);
+  assert(xstrlen((PKMX_WCHAR)u"\uFFFF\u0017\u0001\u0002\u0003a") == 2);
```

Comment:
I wonder if we can make these clearer with some #defines? e.g.

```c++
#define U_UC_SENTINEL u"\uFFFF"
#define U_CODE_ANY u"\u0001"
...
#define U_CODE_DEADKEY u"\u0008"

#define C_CODE_ANY(store) U_UC_SENTINEL U_CODE_ANY store
...
#define C_CODE_DEADKEY(deadkey) U_UC_SENTINEL U_CODE_DEADKEY deadkey

assert(xstrlen((PKMX_WCHAR) C_CODE_DEADKEY("\u0001") u"a") == 2);
```

Then we just need to verify those definitions once and all the tests should become (a) much easier to read, and (b) much less likely to have typos. (Of course, we still need to write the intentionally malformed test strings very carefully!)

Reply:
I changed it for the new methods. For the existing tests in this file I'll do it in a separate PR. I also added a define for the surrogate pair, but I'm not sure if that makes things clearer or hides the fact that it's two bytes.

Reply:
Other tests are done in #6002